File: app/src/main/java/org/deiverbum/app/workers/TodayWorker.java
package org.deiverbum.app.workers;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.hilt.work.HiltWorker;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.deiverbum.app.data.db.dao.TodayDao;
import org.deiverbum.app.data.source.remote.network.ApiService;
import org.deiverbum.app.model.crud.Crud;

import javax.inject.Inject;

import dagger.assisted.Assisted;
import dagger.assisted.AssistedInject;
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.observers.DisposableSingleObserver;
import io.reactivex.rxjava3.schedulers.Schedulers;

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */

@HiltWorker
public class TodayWorker extends Worker {

    @Inject
    ApiService apiService;

    @Inject
    TodayDao mTodayDao;

    @AssistedInject
    public TodayWorker(
            @Assisted @NonNull Context context,
            @Assisted @NonNull WorkerParameters params

    ) {
        super(context, params);
    }

    @NonNull
    @Override
    public Result doWork() {
        try {
            loadCrud();
            return Result.success();
        } catch (Throwable e) {
            return Result.failure();
        }
    }

    public void loadCrud() {
        apiService.postCrud(mTodayDao.getAllSyncStatus())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new DisposableSingleObserver<Crud>() {

                    @Override
                    public void onStart() {
                    }

                    @Override
                    public void onSuccess(Crud r) {
                        try {
                            if (r.haveData) {
                                r.doCrud(mTodayDao);
                                mTodayDao.syncUpdate(r.lastUpdate);
                            }
                        } catch (Exception e) {
                            Log.d("ERR", e.getMessage());
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.d("ERR", e.getMessage() == null ? "" : e.getMessage());
                    }
                });
    }
}

File: app/src/main/java/org/deiverbum/app/workers/TodayWorker.kt
package org.deiverbum.app.workers

import android.content.Context
import androidx.hilt.work.HiltWorker
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import dagger.assisted.Assisted
import dagger.assisted.AssistedInject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.deiverbum.app.core.database.dao.TodayDao
import org.deiverbum.app.core.network.api.TodayApi

/**
 * Esta clase se ocupa de manejar el Worker de sincronización.
 *
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
@HiltWorker
class TodayWorker @AssistedInject constructor(
    @Assisted context: Context,
    @Assisted params: WorkerParameters,
    private val apiService: TodayApi,
    private val todayDao: TodayDao

) : CoroutineWorker(context, params) {


    override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
        return@withContext try {
            loadCrud()
            Result.success()
        } catch (error: Throwable) {
            //TODO quitar println
            println("TWError $error")
            Result.failure()
        }
    }

    /**
     * Este método busca si hay sincronizaciones disponibles en el servidor
     * desde la última sincronización.
     * Si las hay, actualizará esos datos llamando al método [`Crud.doCrud()`](org.deiverbum.app.core.model.data.crud.Crud.doCrud)
     * y actualizará también la fecha de la última sincronización, que será la actual,
     * en la base de datos.
     *
     */
    private suspend fun loadCrud() {
        val crud = apiService.postCrud(todayDao.allSyncStatus)
        if (crud.haveData) {
            crud.doCrud(todayDao)
            if (crud.lastUpdate != "") {
                todayDao.syncUpdate(crud.lastUpdate)
            }
        }
    }
}